### PR TITLE
Update to support databases with authentication

### PR DESCRIPTION
**Pull request by gwbischof**
If the user account is read only, index creation fails. So in this PR indexes creation is changed so that they are made only when needed.

### CODE REVIEW

**Comment by gwbischof** [comment]
Still a draft

File: databroker/assets/base_registry.py [hunk 1]
```diff
@@ -463,6 +463,8 @@ def register_resource(self, spec, root, rpath, rkwargs,
             The uid of the created resource.
 
         '''
+        if callable(getattr(self, '_create_resource_index', None)):
```
Comment (danielballan) [inline databroker/assets/base_registry.py]: If this exists but is not a callable, what does that mean? Should it pass silently?

**Comment by danielballan** [comment]
With fresh eyes I now understand what looks off about this to me. I think this is doing:

```py
In [2]: class A: 
   ...:     def some_method(self): 
   ...:         print('some method!') 
   ...:         if hasattr(self, '_extra_step'): 
   ...:             self._extra_step() 
   ...:                                                                                                                                                                                       

In [3]: class B(A): 
   ...:     def _extra_step(self): 
   ...:         print('extra step!') 
   ...:                                                                                                                                                                                       

In [4]: a = A()                                                                                                                                                                               

In [5]: b = B()                                                                                                                                                                               

In [6]: a.some_method()                                                                                                                                                                       
some method!

In [7]: b.some_method()                                                                                                                                                                       
some method!
extra step!
```

where it would be better to rely on `super()` as in

```py
In [14]: class A: 
    ...:     def some_method(self): 
    ...:         print('some method!') 
    ...:                                                                                                                                                                                      

In [15]: class B(A): 
    ...:     def _extra_step(self): 
    ...:         print('extra step!') 
    ...:     def some_method(self): 
    ...:         super().some_method() 
    ...:         self._extra_step() 
    ...:                                                                                                                                                                                      

In [16]: a = A()                                                                                                                                                                              

In [17]: b = B()                                                                                                                                                                              

In [18]: a.some_method()                                                                                                                                                                      
some method!

In [19]: b.some_method()                                                                                                                                                                      
some method!
extra step!
```

or, equivalently

```py
In [20]: class B(A): 
    ...:     def some_method(self): 
    ...:         super().some_method() 
    ...:         print('extra step!') 
    ...:                                                                                                                                                                                      
```

By using `super()`, we can put all the special logic in the subclass, rather than teaching the base class to anticipate and delegate to any subclass-specific methods that may exist.

File: databroker/assets/base_registry.py (hunk 1 above)
Comment (tacaswell) [reply databroker/assets/base_registry.py]: I agree, `hasattr` or `... is not None` is what I would have naively expected to be here.

**Comment by tacaswell** [comment]
I agree with @danielballan about using `super()` and extending the sub-classes.

Setting up the indexes on init is something we put in early because without the indexes some of our standard operations are super slow and we were working around insufficient devops on the floor.

I think it is ok to either silently eat this exception if we can't create the index or just drop it all together and assert that the devops need to be done correctly.

**Comment by danielballan** [comment]
One random failure on one build, due to known issue #431

**Pull request merged**